crates/wasi-common/src/sys/mod.rs

```diff
@@ -20,3 +23,65 @@ cfg_if! {
 
 pub(crate) use sys_impl::path;
 pub(crate) use sys_impl::poll;
+
+use super::handle::Handle;
+use crate::wasi::types;
+use osdir::OsDir;
+use osfile::OsFile;
+use osother::OsOther;
+use std::convert::TryFrom;
+use std::fs::File;
+use std::io;
+use std::mem::ManuallyDrop;
+use stdio::{Stderr, Stdin, Stdout};
+use sys_impl::get_file_type;
+
+pub(crate) trait AsFile {
+    fn as_file(&self) -> io::Result<ManuallyDrop<File>>;
+}
```

Comment:
Hmm, that's an interesting suggestion. So previously, `AsFile` wasn't part of `Handle` trait since it wasn't fallible, and hence, it wouldn't make much sense for `VirtualFile`, etc. But now, given that we return `Result` I guess we could in principle return `Errno::Badf` when called by a virtual handle? Or should be outright panic in such a case?

Reply:
Oh, right. I wasn't thinking about cases like `VirtualHandle`. I think it's fine as is then.